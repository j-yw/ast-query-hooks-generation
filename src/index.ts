import * as t from "@babel/types";
interface HookParams {
  interfaceName: string;
  hookName: string;
  requestType: string;
  responseType: string;
  queryServiceMethod: string;
  queryKey: string;
}

function createReturnType(interfaceName: string) {
  return t.tsTypeReference(
    t.identifier(interfaceName),
    t.tsTypeParameterInstantiation([t.tsTypeReference(t.identifier("TData"))])
  );
}

export function generateHookAST({
  interfaceName,
  hookName,
  requestType,
  responseType,
  queryServiceMethod,
  queryKey,
}: HookParams) {
  const useHookFunction = t.arrowFunctionExpression(
    // params
    [
      t.objectPattern([
        t.objectProperty(t.identifier("request"), t.identifier("request")),
        t.objectProperty(t.identifier("options"), t.identifier("options")),
      ]),
    ],

    // body
    t.blockStatement([
      t.returnStatement(
        t.callExpression(t.identifier("useQuery"), [
          t.arrayExpression([
            t.stringLiteral(queryKey),
            t.identifier("request"),
          ]),
          t.arrowFunctionExpression(
            // params
            [],
            // tsInterfaceBody
            t.blockStatement([
              t.ifStatement(
                t.unaryExpression("!", t.identifier("queryService")),
                t.throwStatement(
                  t.newExpression(t.identifier("Error"), [
                    t.stringLiteral("queryService is not defined"),
                  ])
                )
              ),
              t.returnStatement(
                t.callExpression(
                  t.memberExpression(
                    t.identifier("queryService"),
                    t.identifier(queryServiceMethod)
                  ),
                  [t.identifier("request")]
                )
              ),
            ]),
            // async?
            false
          ),
          t.identifier("options"),
        ])
      ),
    ]),
    //Async?
    false
  );

  useHookFunction.returnType = t.tsTypeAnnotation(
    createReturnType(interfaceName)
  );

  const interfaceAST = t.tsInterfaceDeclaration(
    t.identifier(interfaceName),

    t.tsTypeParameterDeclaration([t.tsTypeParameter(null, null, "TData")]),

    [
      t.tsExpressionWithTypeArguments(
        t.identifier("ReactQueryParams"),
        t.tsTypeParameterInstantiation([
          t.tsTypeReference(t.identifier(responseType)),
          t.tsTypeReference(t.identifier("TData")),
        ])
      ),
    ],

    t.tsInterfaceBody([
      t.tsPropertySignature(
        t.identifier("request"),
        t.tsTypeAnnotation(
          t.tsUnionType([
            t.tsTypeReference(t.identifier(requestType)),
            // t.tsUndefinedKeyword(),
          ])
        )
      ),
    ])
  );

  const hookFunctionX = t.variableDeclaration("const", [
    t.variableDeclarator(t.identifier(hookName), useHookFunction),
  ]);

  const arrowFunction = hookFunctionX.declarations[0]
    .init as t.ArrowFunctionExpression;
  const blockStatement = arrowFunction.body as t.BlockStatement;
  const returnStatement = blockStatement.body[0] as t.ReturnStatement;
  const useQueryCall = returnStatement.argument as t.CallExpression;

  if (t.isCallExpression(useQueryCall)) {
    useQueryCall.typeParameters = t.tsTypeParameterInstantiation([
      t.tsTypeReference(t.identifier(responseType)),
      t.tsTypeReference(t.identifier("Error")),
      t.tsTypeReference(t.identifier("TData")),
    ]);
  } else {
    console.error("Expected useQuery call not found in the generated AST");
  }

  return t.program([t.exportNamedDeclaration(interfaceAST), hookFunctionX]);
}
