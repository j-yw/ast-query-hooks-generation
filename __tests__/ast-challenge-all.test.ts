import { generateCustomHook } from "../src/index";
import generate from "@babel/generator";
import * as prettier from "prettier";
import * as fs from "fs";
import * as path from "path";

interface MethodData {
  requestType: string;
  responseType: string;
}

interface Methods {
  [methodName: string]: MethodData;
}

describe("Generate multiple hooks: ", () => {
  it("worked", async () => {
    const methods: Methods = JSON.parse(
      fs.readFileSync(path.join(__dirname, "../example-methods.json"), "utf-8")
    );

    for (const [methodName, methodData] of Object.entries(methods)) {
      const ast = generateCustomHook({
        interfaceName: `Use${methodName}Query`,
        customHookName: `use${methodName}`,
        requestType: methodData.requestType,
        responseType: methodData.responseType,
        queryServiceMethod: methodName.toLowerCase(),
        queryKey: `${
          methodName.charAt(0).toLowerCase() + methodName.slice(1)
        }Query`,
      });

      const { code } = generate(ast as any);
      const formattedCode = prettier.format(code, { parser: "typescript" });
      console.log(formattedCode);
      expect(formattedCode).toMatchSnapshot(methodName);
    }
  });
});
