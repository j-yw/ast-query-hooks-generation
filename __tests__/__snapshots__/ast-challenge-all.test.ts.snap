// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate multiple hooks:  should return multiple react query hooks: EstimateSwapExactAmountIn 1`] = `
"export interface UseEstimateSwapExactAmountInQuery<TData>
  extends ReactQueryParams<QuerySwapExactAmountInResponse, TData> {
  request?: QuerySwapExactAmountInRequest;
}

const useEstimateSwapExactAmountIn = <TData = QuerySwapExactAmountInResponse>({
  request,
  options,
}: UseEstimateSwapExactAmountInQuery<TData>) => {
  return useQuery<QuerySwapExactAmountInResponse, Error, TData>(
    [\\"estimateswapexactamountinQuery\\", request],
    () => {
      if (!queryService) throw new Error(\\"Query Service not initialized\\");
      return queryService.estimateswapexactamountin(request);
    },
    options
  );
};
"
`;

exports[`Generate multiple hooks:  should return multiple react query hooks: EstimateSwapExactAmountOut 1`] = `
"export interface UseEstimateSwapExactAmountOutQuery<TData>
  extends ReactQueryParams<QuerySwapExactAmountOutResponse, TData> {
  request?: QuerySwapExactAmountOutRequest;
}

const useEstimateSwapExactAmountOut = <
  TData = QuerySwapExactAmountOutResponse
>({
  request,
  options,
}: UseEstimateSwapExactAmountOutQuery<TData>) => {
  return useQuery<QuerySwapExactAmountOutResponse, Error, TData>(
    [\\"estimateswapexactamountoutQuery\\", request],
    () => {
      if (!queryService) throw new Error(\\"Query Service not initialized\\");
      return queryService.estimateswapexactamountout(request);
    },
    options
  );
};
"
`;

exports[`Generate multiple hooks:  should return multiple react query hooks: NumPools 1`] = `
"export interface UseNumPoolsQuery<TData>
  extends ReactQueryParams<QueryNumPoolsResponse, TData> {
  request?: QueryNumPoolsRequest;
}

const useNumPools = <TData = QueryNumPoolsResponse>({
  request,
  options,
}: UseNumPoolsQuery<TData>) => {
  return useQuery<QueryNumPoolsResponse, Error, TData>(
    [\\"numpoolsQuery\\", request],
    () => {
      if (!queryService) throw new Error(\\"Query Service not initialized\\");
      return queryService.numpools(request);
    },
    options
  );
};
"
`;

exports[`Generate multiple hooks:  should return multiple react query hooks: Pool 1`] = `
"export interface UsePoolQuery<TData>
  extends ReactQueryParams<QueryPoolResponse, TData> {
  request?: QueryPoolRequest;
}

const usePool = <TData = QueryPoolResponse>({
  request,
  options,
}: UsePoolQuery<TData>) => {
  return useQuery<QueryPoolResponse, Error, TData>(
    [\\"poolQuery\\", request],
    () => {
      if (!queryService) throw new Error(\\"Query Service not initialized\\");
      return queryService.pool(request);
    },
    options
  );
};
"
`;

exports[`Generate multiple hooks:  should return multiple react query hooks: PoolParams 1`] = `
"export interface UsePoolParamsQuery<TData>
  extends ReactQueryParams<QueryPoolParamsResponse, TData> {
  request?: QueryPoolParamsRequest;
}

const usePoolParams = <TData = QueryPoolParamsResponse>({
  request,
  options,
}: UsePoolParamsQuery<TData>) => {
  return useQuery<QueryPoolParamsResponse, Error, TData>(
    [\\"poolparamsQuery\\", request],
    () => {
      if (!queryService) throw new Error(\\"Query Service not initialized\\");
      return queryService.poolparams(request);
    },
    options
  );
};
"
`;

exports[`Generate multiple hooks:  should return multiple react query hooks: Pools 1`] = `
"export interface UsePoolsQuery<TData>
  extends ReactQueryParams<QueryPoolsResponse, TData> {
  request?: QueryPoolsRequest;
}

const usePools = <TData = QueryPoolsResponse>({
  request,
  options,
}: UsePoolsQuery<TData>) => {
  return useQuery<QueryPoolsResponse, Error, TData>(
    [\\"poolsQuery\\", request],
    () => {
      if (!queryService) throw new Error(\\"Query Service not initialized\\");
      return queryService.pools(request);
    },
    options
  );
};
"
`;

exports[`Generate multiple hooks:  should return multiple react query hooks: SpotPrice 1`] = `
"export interface UseSpotPriceQuery<TData>
  extends ReactQueryParams<QuerySpotPriceResponse, TData> {
  request?: QuerySpotPriceRequest;
}

const useSpotPrice = <TData = QuerySpotPriceResponse>({
  request,
  options,
}: UseSpotPriceQuery<TData>) => {
  return useQuery<QuerySpotPriceResponse, Error, TData>(
    [\\"spotpriceQuery\\", request],
    () => {
      if (!queryService) throw new Error(\\"Query Service not initialized\\");
      return queryService.spotprice(request);
    },
    options
  );
};
"
`;

exports[`Generate multiple hooks:  should return multiple react query hooks: TotalLiquidity 1`] = `
"export interface UseTotalLiquidityQuery<TData>
  extends ReactQueryParams<QueryTotalLiquidityResponse, TData> {
  request?: QueryTotalLiquidityRequest;
}

const useTotalLiquidity = <TData = QueryTotalLiquidityResponse>({
  request,
  options,
}: UseTotalLiquidityQuery<TData>) => {
  return useQuery<QueryTotalLiquidityResponse, Error, TData>(
    [\\"totalliquidityQuery\\", request],
    () => {
      if (!queryService) throw new Error(\\"Query Service not initialized\\");
      return queryService.totalliquidity(request);
    },
    options
  );
};
"
`;

exports[`Generate multiple hooks:  should return multiple react query hooks: TotalPoolLiquidity 1`] = `
"export interface UseTotalPoolLiquidityQuery<TData>
  extends ReactQueryParams<QueryTotalPoolLiquidityResponse, TData> {
  request?: QueryTotalPoolLiquidityRequest;
}

const useTotalPoolLiquidity = <TData = QueryTotalPoolLiquidityResponse>({
  request,
  options,
}: UseTotalPoolLiquidityQuery<TData>) => {
  return useQuery<QueryTotalPoolLiquidityResponse, Error, TData>(
    [\\"totalpoolliquidityQuery\\", request],
    () => {
      if (!queryService) throw new Error(\\"Query Service not initialized\\");
      return queryService.totalpoolliquidity(request);
    },
    options
  );
};
"
`;

exports[`Generate multiple hooks:  should return multiple react query hooks: TotalShares 1`] = `
"export interface UseTotalSharesQuery<TData>
  extends ReactQueryParams<QueryTotalSharesResponse, TData> {
  request?: QueryTotalSharesRequest;
}

const useTotalShares = <TData = QueryTotalSharesResponse>({
  request,
  options,
}: UseTotalSharesQuery<TData>) => {
  return useQuery<QueryTotalSharesResponse, Error, TData>(
    [\\"totalsharesQuery\\", request],
    () => {
      if (!queryService) throw new Error(\\"Query Service not initialized\\");
      return queryService.totalshares(request);
    },
    options
  );
};
"
`;
